qiskit_experiments/framework/backend_timing_mixin.py

```diff
@@ -0,0 +1,169 @@
+        a subsequent instruction.
+
+        This method is useful when it is desired that a delay instruction
+        accurately reflect how long the hardware will pause between two other
+        instructions.
+
+        Args:
+            time: The nominal delay time to convert
+            next_instruction: Either "pulse", "acquire", or "both" to indicate
+                whether the next instruction will be a pulse, an acquire
+                instruction, or both.
+
+        Returns:
+            The delay duration in samples or seconds, depending on the value of
+            :meth:`.BackendTimingMixin._delay_unit`.
+
+        Raises:
+            QiskitError: Bad value for ``next_instruction``.
+        """
+        if self._delay_unit == "s":
+            return time
+
+        timing_constraints = getattr(self.backend.configuration(), "timing_constraints", {})
+        pulse_alignment = timing_constraints.get("pulse_alignment", 1)
+        acquire_alignment = timing_constraints.get("acquire_alignment", 1)
+
+        if next_instruction == "both":
+            # We round the delay values to the least common multiple of the
+            # alignment constraints so that the delay can fill all the time
+            # until a subsequent pulse satisfying the alignment constraints.
+            #
+            # Replace with math.lcm(pulse_alignment, acquire_alignment) when
+            # dropping support for Python 3.8
+            granularity = (
+                pulse_alignment * acquire_alignment // math.gcd(pulse_alignment, acquire_alignment)
+            )
+        elif next_instruction == "pulse":
+            granularity = pulse_alignment
+        elif next_instruction == "acquire":
+            granularity = acquire_alignment
+        else:
+            raise QiskitError(f"Bad value for next_instruction: {next_instruction}")
+
+        samples = int(round(time / self._dt / granularity) * granularity)
+
+        return samples
+
+    def _pulse_duration(self, time: float) -> int:
+        """The number of samples giving a valid pulse duration closest to ``time``
+
+        Args:
+            time: Pulse duration in seconds
+
+        Returns:
+            The number of samples corresponding to ``time``
+
+        Raises:
+            QiskitError: The backend timing constraints' min_length is not a
+                multiple of granularity
+        """
+        timing_constraints = getattr(self.backend.configuration(), "timing_constraints", {})
+        granularity = timing_constraints.get("granularity", 1)
+        min_length = timing_constraints.get("min_length", 1)
+
+        samples = int(round(time / self._dt / granularity)) * granularity
+        samples = max(samples, min_length)
+
+        if min_length % granularity != 0:
+            raise QiskitError("Backend timing does not match assumptions!")
+
+        return samples
+
+    def _delay_time(self, time: float, next_instruction: str = "pulse") -> float:
```

Comment:
I'm not sure if these -`time` methods are really needed, i.e. in the most cases

```python
# this is a particular example of T1

for delay in experiment_options.delays:
    delay_dt = self._delay_duration(delay, "acquire")
   ...
    assigned_circuit = circuit.assign_parameters({p_delay: delay_dt}, inplace=False)
    assigned_circuit.metadata = {"xdata": delay_dt * self._dt}
```

it's simpler to directly convert dt into SI rather than calling the conversion method twice. Am I missing some edge case? For example, you can return `self._dt = 1` when dt is not defined in that backend, i.e. consider the value is in SI unit.

Reply:
At the moment, I am raising an exception if `dt` is not available rather than defaulting to 1. The question is what does `dt` represent. If it is "the time in seconds in between samples of pulses" then there is appropriate default value and using 1 could lead to unexpected results. If instead `dt` represents "the number that you multiply pulse delay durations by to get seconds" having 1 as the default could work for the case where delays are set in seconds instead of samples.

With how I have things now, when `dt` is not available, the `_delay_unit` is "s" and the `_delay_duration` is in seconds, just like `_delay_time`. So `_delay_time` gives you time always and is convenient when you want the time value for analysis (like a T1 measurement) while `_delay_duration` gives you the value to pass to the `duration` argument of `QuantumCircuit.delay`. For the pulse cases, it is true that the `_time` method is just mulitplying by `dt`. I included that one just to mirror the delay method. I also think it is marginally more readable to have a function that gives you time than to mulitply by a `dt` coefficient.

Reply:
Good point. I didn't think about the definition carefully. These description would be mathematically same, however "the time in seconds in between samples of pulses" sounds reasonable to me. In this sense, your approach is correct. Probably the definition of dt and usage of the -time methods can be described in the API doc.

Reply:
Marking as resolved though you can make a new comment if you want. Now there are _samples and _time methods. You can see if you think the time methods are worth keeping or not. They are basically just multiplying by dt but may or may not make things more readable.